frontend/public/components/monitoring/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,493 @@
+   * }
+   */
+  const normalizeRouteLabels = (returnRegexMatches = false) => {
+    const matches = _.filter(routeLabels, (routeLabel) =>
+      returnRegexMatches ? routeLabel.isRegex : !routeLabel.isRegex,
+    );
+    const matchObj = _.map(matches, (label) => {
+      return { [label.name]: label.value };
+    });
+    return _.assign({}, ...matchObj);
+  };
+
+  /**
+   * Returns new Route object
+   * Ex:
+   * {
+   *   receiver: myNewReceiver,
+   *   match: {
+   *     "severity": "warning",
+   *     "cluster": "myCluster"
+   *   }
+   *   match_re {
+   *     "service": "^(foo1|foo2|baz)$",
+   *   }
+   * }
+   */
+  const createRoute = (receiver: AlertManagerReceiver) => {
+    const matchLabels = normalizeRouteLabels();
+    const matchReLabels = normalizeRouteLabels(true);
+
+    // construct new route object
+    const newRoute = { receiver: receiver.name };
+    if (!_.isEmpty(matchLabels)) {
+      _.set(newRoute, 'match', matchLabels);
+    }
+    if (!_.isEmpty(matchReLabels)) {
+      _.set(newRoute, 'match_re', matchReLabels);
+    }
+    return newRoute;
+  };
+
+  /**
+   * Returns new Receiver object
+   * Ex:
+   * {
+   *   name: MyNewReceiver
+   *   pagerduty_configs: {
+   *     routing_key: <integration_key>
+   *   }
+   * }
+   */
+  const createReceiver = () => {
+    const receiverConfigKey = `${receiverType}_configs`;
+    let receiverConfig;
+    switch (receiverType) {
+      case 'pagerduty':
+        // eslint-disable-next-line no-case-declarations
+        const pagerDutyIntegrationKeyName = `${
+          pagerDutyValues.integrationType === 'events' ? 'routing' : 'service'
+        }_key`;
+        receiverConfig = {
+          [pagerDutyIntegrationKeyName]: pagerDutyValues.integrationKey,
+        };
+        break;
+      case 'webhook':
+        receiverConfig = { url: webhookUrl };
+        break;
+      default:
+    }
+    return {
+      name: receiverName,
+      [receiverConfigKey]: [{ ...receiverConfig }],
+    };
+  };
+
+  const save = (e) => {
+    e.preventDefault();
+
+    // add new receiver to alert manager receivers
+    const newReceiver = createReceiver();
+    const receivers = _.get(config, 'receivers', []);
+    receivers.push(newReceiver);
+    _.set(config, 'receivers', receivers);
+
+    // set as default or create new route
+    if (saveAsDefaultReceiver) {
+      _.set(route, 'receiver', newReceiver.name);
+    } else if (!_.isEmpty(routeLabels)) {
+      const routes = _.get(route, 'routes', []);
+      const newRoute = createRoute(newReceiver);
+      routes.push(newRoute);
+      _.set(route, 'routes', routes);
+    }
+
+    setInProgress(true);
+    patchAlertManagerConfig(secret, config).then(
+      () => {
+        setErrorMsg('');
+        setInProgress(false);
+        history.push('/monitoring/alertmanagerconfig');
+      },
+      (err) => {
+        setErrorMsg(err.message);
+        setInProgress(false);
+      },
+    );
+  };
+
+  const receiverTypes = {
```

Comment:
Probably better to move simple constants outside the component.

Reply:
I've done this in a yet-to-be-pushed PR